relays/substrate/src/headers_initialize.rs

```diff
@@ -88,50 +83,153 @@ async fn do_initialize<SourceChain: Chain, TargetChain: Chain>(
-
-	let initial_header = match raw_initial_header {
-		Some(raw_initial_header) => SourceChain::Header::decode(&mut &raw_initial_header.0[..])
-			.map_err(|err| format!("Failed to decode {} initial header: {:?}", SourceChain::NAME, err))?,
-		None => source_client
-			.header_by_hash(source_genesis_hash)
-			.await
-			.map_err(|err| format!("Failed to retrive {} genesis header: {:?}", SourceChain::NAME, err))?,
-	};
-
-	let raw_initial_authorities_set = match raw_initial_authorities_set {
-		Some(raw_initial_authorities_set) => raw_initial_authorities_set.0,
-		None => source_client
-			.grandpa_authorities_set(source_genesis_hash)
-			.await
-			.map_err(|err| {
-				format!(
-					"Failed to retrive {} authorities set at genesis header: {:?}",
-					SourceChain::NAME,
-					err
-				)
-			})?,
-	};
-	let initial_authorities_set =
-		GrandpaAuthoritiesSet::decode(&mut &raw_initial_authorities_set[..]).map_err(|err| {
-			format!(
-				"Failed to decode {} initial authorities set: {:?}",
+	// In ideal world we just need to get best finalized header and then to read GRANDPA authorities
+	// set (`pallet_grandpa::CurrentSetId` + `GrandpaApi::grandpa_authorities()`) at this header.
+	//
+	// But now there are problems with this approach - `CurrentSetId` may return invalid value. So here
+	// we're waiting for the next justification, read the authorities set and then try to figure out
+	// the set id with bruteforce.
+	let mut justifications = source_client
+		.subscribe_justifications()
+		.await
+		.map_err(|err| format!("Failed to subscribe to {} justifications: {:?}", SourceChain::NAME, err))?;
+
+	// Read next justification - the header that it finalizes will be used as initial header.
+	let justification = justifications.next().await.ok_or_else(|| {
+		format!(
+			"Failed to read {} justification from the stream: stream has ended unexpectedly",
+			SourceChain::NAME,
+		)
+	})?;
+
+	// Read initial header.
+	let (initial_header_hash, initial_header_number) =
+		decode_justification_target::<SourceChain::Header>(&justification.0)
+			.map_err(|err| format!("Failed to decode {} justification: {:?}", SourceChain::NAME, err))?;
+	let initial_header = source_header(&source_client, initial_header_hash).await?;
+	log::trace!(target: "bridge", "Selected {} initial header: {}/{}",
+		SourceChain::NAME,
+		initial_header_number,
+		initial_header_hash,
+	);
+
+	// Read GRANDPA authorities set at initial header.
+	let initial_authorities_set = source_authorities_set(&source_client, initial_header_hash).await?;
+	log::trace!(target: "bridge", "Selected {} initial authorities set: {:?}",
+		SourceChain::NAME,
+		initial_authorities_set,
+	);
+
+	// If initial header changes the GRANDPA authorities set, then we need previous authorities
+	// to verify justification.
+	let mut authorities_for_verification = initial_authorities_set.clone();
+	let schedules_change = find_grandpa_authorities_scheduled_change(&initial_header).is_some();
```

Comment:
Do you think it would be worth it to check for forced changes?

If the initial header schedules forced changes with a delay of `N` blocks the finality pallet has no way to knowing that header `initial + N` will enact a new authority set change and I think we'll get stuck